### DISCUSSION TITLE
Display water pressure in home page

### DISCUSSION THREAD

**Discussion by PlaneStuffRC** [Dan-in-CA/SIP#266] [Ideas]
I've been playing around with the with the pump control plugin, and i believe its useful to display the water pressure and some other parameters on the home page. Any hints on how to do that ?

**Comment by Dan-in-CA** [comment]
Take a look at the "plugin data" branch of this repo. It has a couple of ways to display that sort of data.

**Comment by PlaneStuffRC** [comment]
Tried to use the proto plugin from plugin_data branch (replaced webpages.py also) but i miss something because its not working.


`
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/pi/SIP/plugins/proto.py", line 50, in footer_test
    example = showInFooter() #  instantiate class to enable data in footer
  File "/home/pi/SIP/webpages.py", line 768, in __init__
    self._idx = len(gv.pluginFtr)
AttributeError: 'module' object has no attribute 'pluginFtr'
`

**Comment by Dan-in-CA** [comment]
Hi Adrian,
I cloned a fresh copy of SIP, changed to the plugin_data branch from the SIP directory with:
`git checkout plugin_data`
Opened the plugin manager and enabled the proto plugin and the data display appeared below the timeline on the home screen. I tried it with both Python 2 and 3.

There is another data display that can appear on the timeline next to the remaining time for the running station. This requires an irrigation program to be set up and run. You can use the "run now" button on the programs page for testing.

Dan

**Comment by PlaneStuffRC** [comment]
Hi Dan,

I've tried this way and it works, the only issue i found its the version:  `Software version: 4.0.27 (2020-04-20)` and the update plugin doesn't work. Basically it would be great if we can use the proto plugin from Plugin_data with the latest rev of SIP.

Adrian.

**Comment by Dan-in-CA** [comment]
Yes, There have been many updates to the master branch after the plugin was posted.
I will work on updating the plugin_data branch. 
It may take a few days and I will post a message here as soon as it is ready.

Dan

**Comment by PlaneStuffRC** [comment]
Hi Dan,

So basically what I'm trying to do, is to make a custom pump_controller plugin and display some data on the home page. 
For that i modified the original pump_control.py to display the data in the footer.
The only issue i see with this one is that, even if the Pump Status changes, i have to refresh the page to updated the value, the upper one works as it should.
![pc_browser](https://user-images.githubusercontent.com/15807238/121347613-1140ba80-c930-11eb-88b2-196b17d8f4a5.png)

This is the mobile version:
![mobile_browser](https://user-images.githubusercontent.com/15807238/121347651-1aca2280-c930-11eb-84d7-1175d4f22517.jpeg)


    def  footer_test():
    
    """
    
    Example of plugin data display in UI footer.
    
    Run in background thread.
    
    """
    
    water_pressure = showInFooter() # instantiate class to enable data in footer
    
    water_pressure.label =  u"Pressure"
    
    water_pressure.val =  0
    
    water_pressure.unit =  u" mBar"
    
      
    
    pump_status = showInFooter() # instantiate class to enable data in footer
    
    pump_status.label =  u"Pump Status"
    
    pump_status.val =  0
    
    pump_status.unit =  u" "
    
      
    
    while  True: # Simulate plugin data
    
    water_pressure.val = get_now_pressure() # update plugin data
    
    pump_status.val = get_now_status() # update plugin data
    
    print('Pressure',water_pressure)
    
    print('Status',pump_status)
    
    sleep(2)

**Comment by Dan-in-CA** [comment]
I am not getting the images.

It looks like both values should update. Is the get_now_status() function returning updated values?
Do the print functions show updated values?

What happens if you try to show the status only? does it update then?

**Comment by PlaneStuffRC** [comment]
Solved images issue.
The get_now_status() its working properly because if i show only the status in the footer, it works. 
Also the print functions are displaying updated values.

**Comment by Dan-in-CA** [comment]
I am getting close to a fix.
the code that updates the footer with the plugin data is in base.html starting at about line 157:

             function dataTimer() {
                // Call to get plugin data
                jQuery.getJSON("/api/plugins", function(piVals) {
                	var parID = 'piData'  + piVals.fdata[0][0] 
                	jQuery('p#'+parID ).find('span[name="val"]').text(piVals.fdata[0][1])
                } 

It only updates one paragraph in the display.  I will be changing it to update all paragraphs. I am also making some other changes to the related code.

It may be a day or two until it is ready.

**Comment by Dan-in-CA** [comment]
Adrian,
I made a new branch named "PID_footer"on the SIP repo  (PID = Plug In Data) .

It is a branch of the current SIP rev with the plugin data displayed in the footer. there are 2 example data sets in the Proto plugin that both update without needing a screen refresh. It should work with any number of data sets.

I also tweaked the CSS to make it responsive  so it works better on mobile devices.
Let me know what you think.
Dan

**Comment by PlaneStuffRC** [comment]
Hi Dan,

Everything its working now and the mobile page its looking much better.

![WhatsApp Image 2021-06-11 at 11 14 00 PM](https://user-images.githubusercontent.com/15807238/121743964-d8107200-cb0a-11eb-9ebb-3b4791902d16.jpeg)

**Comment by PlaneStuffRC** [comment]
To explain more on what I'm trying to achieve:
![block_diag](https://user-images.githubusercontent.com/15807238/121745020-78b36180-cb0c-11eb-972c-6f5adebd9e65.png)
 
Basically I'm writing a new plugin from the original one, that will have can-bus communication:
Example, very easy to do it in python with soketcan.
`        if water_pressure.val <= 50:`
`            msg = can.Message(arbitration_id=0x00000066, data=[ 0x22, 0x17, 0x00, 0, 0, 0, 0, 0x39], is_extended_id=True)`
`            bus.send(msg)`
`        else:`
`            msg = can.Message(arbitration_id=0x00000066, data=[ 0x22, 0x17, 0x64, 0, 0, 0, 0, 0x9D], is_extended_id=True)`
`            bus.send(msg)`
Regarding the module, I'll be using Arduino with MCP2515 for simplicity and low cost.
The plugin web page will need some changes because i will be adding some new features, like a flow sensor and a wait time after reaching a low water level.

I will post the project in a few days--weeks :)

**Comment by PlaneStuffRC** [comment]
> 
> 
> Adrian,
> I made a new branch named "PID_footer"on the SIP repo (PID = Plug In Data) .
> 
> It is a branch of the current SIP rev with the plugin data displayed in the footer. there are 2 example data sets in the Proto plugin that both update without needing a screen refresh. It should work with any number of data sets.
> 
> I also tweaked the CSS to make it responsive so it works better on mobile devices.
> Let me know what you think.
> Dan

Hi Dan,

I was thinking, maybe you can add the showOnTimeline functionality to "PID_footer"on the SIP repo (PID = Plug In Data) .
Because, i can add flow sensor data to each station from this plugin https://github.com/DaveSprague/sip_plugins/tree/master/flow_sensors.

**Comment by Dan-in-CA** [comment]
Yes I was already planning to add that. Just wanted to be sure the footer part was working properly.
After adding the timeline display functionality I plan to merge the PID_footer branch into the master branch.

There is also a part of the plugin data functionality that can add plugin data to SIP's run log. that needs a bit more testing.

**Comment by Dan-in-CA** [comment]
OK. I am moving this issue to Discussions in the Ideas section.

**Reply by Dan-in-CA** [reply]
I added plugin data capability for the timeline to the PID_footer branch.
There are changes to the Proto plugin to make testing easier using simulated plugin data.

Please note that in order to display data on the timeline an irrigation program or a Run-once program must be running.

**Reply by PlaneStuffRC** [reply]
I will test it tonight. Thanks!

**Reply by PlaneStuffRC** [reply]
I tested the new PID_footer and i get no data on the timeline :00:52, : , : 

![timeline_issue](https://user-images.githubusercontent.com/15807238/122113018-3b134900-ce2a-11eb-9700-82195e4fc537.png)

**Reply by Dan-in-CA** [reply]
The commas and colons are there so it is partly working but as you say no values.
I will look into it.

**Reply by Dan-in-CA** [reply]
My local development copy is working properly so I did a new commit and re-pushed to the PID_footer branch.
Testing a fresh clone seems to work. Something must not have gotten included in the previous push.

I am working on getting plugin data  included in SIP's log. If you have a flow sensor you probably would not always be watching the flow data in real time. having a "total used" in the log for each run could be very useful.

**Reply by PlaneStuffRC** [reply]
I think  I'm missing something because I did copy a fresh copy right now and git checkout PID_footer. Im getting the same result. Could be browser related ? Tried on chrome and mozila.

**Reply by Dan-in-CA** [reply]
I also cleared the cache on my browser (Firefox).

**Reply by Dan-in-CA** [reply]
I just tested on Chrome and my Android phone Seems to be working.

**Reply by PlaneStuffRC** [reply]
Tried this also, no change

**Reply by Dan-in-CA** [reply]
Strange. I will keep checking.

**Reply by PlaneStuffRC** [reply]
If you want to test in on my computer i can give you remote control (team viewer )

**Reply by Dan-in-CA** [reply]
Let me do a bit more testing on this end first.
I am wondering what the time difference is between my Northern California USA location and  Bucharest.

**Reply by Dan-in-CA** [reply]
I did a direct clone of the PID_footer branch using:
`git clone -b PID_footer --single-branch https://github.com/Dan-in-CA/SIP.git`

On my development Pi it works fine. 
When I cloned the branch to a different Pi it does not show the  data. Scratching my head on this one.
I will keep working on this.

**Reply by Dan-in-CA** [reply]
When I run SIP under Python 3 the data shows.
`python3 sip.py`
When run under Python2.7 the data does not show. It is a Python version incompatibility. Looking for the source of the problem.

**Reply by PlaneStuffRC** [reply]
> 
> 
> Let me do a bit more testing on this end first.
> I am wondering what the time difference is between my Northern California USA location and Bucharest.

Should be around 10 hr

**Reply by PlaneStuffRC** [reply]
> 
> 
> When I run SIP under Python 3 the data shows.
> `python3 sip.py`
> When run under Python2.7 the data does not show. It is a Python version incompatibility. Looking for the source of the problem.

I tested now with python3 and it works

**Reply by Dan-in-CA** [reply]
Fixed now for both versions of Python.
Update pushed to repo.

**Reply by PlaneStuffRC** [reply]
Tested, and it works!
